eclipse/src/saros/ui/decorators/SharedProjectDecorator.java

```diff
@@ -106,11 +108,15 @@ public void decorate(Object element, IDecoration decoration) {
 
     IResource resource = (IResource) element;
 
-    if (!currentSession.isShared(ResourceAdapterFactory.create(resource))) return;
+    Set<IReferencePoint> referencePoints = session.getReferencePoints();
+    saros.filesystem.IResource resourceWrapper =
```

Comment:
Am I right that the runtime changed from O(1) to O(n) ? It does not hurt that much since this is called in a background thread.

But as some users already pointed out they are sharing up to 40 projects (hopefully not container over 100000 files).

Reply:
Yes, it changed to O(n) where 'n' is the number of shared reference points. As you already mentioned, this should hopefully not be too big of an issue as this is only done in a background thread. Furthermore, I would presume that the decorator is only loaded for resources that are actually displayed in the project view, meaning the computations should not happen all at once (besides the session start) but rather are drawn out while the user is expanding the view.

In general, the resource lookup running in O(#ref-points) is something that I would ideally have liked to avoid, but I could not come up with a way around it.
With the new reference point system, any directory in the workspace could potentially be a reference point. And, with the new resource being defined relative to their reference point, we have to figure out the correct reference point to use before we can create a Saros resource object. The only way to do this I could think of was going through all shared reference points and compare paths.

Reply:
You could to it in O(log n) by changing the the ReferencePointMapper in such a way that it maintain a sorted list that then can be used by a binary search algorithm to find the referencePoint. However the speedup is very hard to measure, especially with just a few project. On the other hand you could introduce a little "cache", e.g always remember the last used reference point and use this reference point as first comparison.

Reply:
From testing, it looks like Eclipse is actually decorating all shared files, not just the shown ones. So an optimization here would really go a long way here.

Adding the least recently used reference point should works in most cases as the basic decorating process works along the tree branches in a depth-first search manner. But I am not sure whether this process is parallelized in any way (i.e. working to all tree branches in parallel), which would make this kind of optimization unusable.

Reply:
Added optimization.